### PR TITLE
RUM-3294 Add the Core Tracer tests

### PR DESCRIPTION
**Pull request by mariusc83**
### What does this PR do?

In this PR we are porting the core unit tests for the Java CoreTracer logic we borrowed from the APM Agent Java project. Please have in mind that in order to have consistency those tests and the logic around are matching exactly their groovy equivalents so no randomness was used in the input data and no forgery was added. 

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)

### CODE REVIEW

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/CoreSpanBuilderTest.kt [hunk 1]
```diff
@@ -0,0 +1,534 @@
...
+import org.junit.jupiter.params.ParameterizedTest
+import org.junit.jupiter.params.provider.Arguments
+import org.junit.jupiter.params.provider.CsvSource
+import org.junit.jupiter.params.provider.MethodSource
+import org.mockito.kotlin.doReturn
+import org.mockito.kotlin.mock
+import java.util.concurrent.TimeUnit
+import java.util.stream.Stream
+
+@Timeout(5)
```
Comment (0xnm) [inline features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/CoreSpanBuilderTest.kt:38]: is this annotation still needed? in which case we can get a timeout?

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/CoreSpanBuilderTest.kt [hunk 2]
```diff
@@ -0,0 +1,534 @@
...
+    override fun cleanup() {
+        super.cleanup()
+        tracer.close()
+    }
+
+    @Test
+    fun `build simple span`() {
+        val span = tracer.buildSpan(instrumentationName, "op name").withServiceName("foo").start()
+        assertThat(span.operationName).isEqualTo("op name")
+    }
```
Comment (0xnm) [inline features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/CoreSpanBuilderTest.kt:61]: Let's split blocks into `Given/When/Then` here and in other tests.

Also we maybe can generate fake data as we do in other tests for the SDK (this is useful for other tests as well).

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/CoreSpanBuilderTest.kt [hunk 3]
```diff
@@ -0,0 +1,534 @@
...
+        tags.forEach {
+            builder = builder.withTag(it.key, it.value)
+        }
+
+        var span = builder.start()
+
+        assertThat(span.operationName).isEqualTo(expectedName)
+        assertThat(span.tags).containsAllEntriesOf(tags)
+
+        span = tracer.buildSpan(instrumentationName, expectedName).withServiceName("foo").start()
```
Comment (0xnm) [inline features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/CoreSpanBuilderTest.kt:84]:
```suggestion
        val firstSpan = builder.start()

        assertThat(firstSpan.operationName).isEqualTo(expectedName)
        assertThat(firstSpan.tags).containsAllEntriesOf(tags)

        val secondSpan = tracer.buildSpan(instrumentationName, expectedName).withServiceName("foo").start()
```

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/CoreSpanBuilderTest.kt [hunk 4]
```diff
@@ -0,0 +1,534 @@
...
+            .withServiceName("foo")
+            .withStartTimestamp(expectedTimestamp)
+            .start()
+
+        // get return nano time
+        assertThat(span.startTime).isEqualTo(expectedTimestamp * 1000L)
+
+        // auto-timestamp in nanoseconds
+        val start = System.currentTimeMillis()
+        span = tracer.buildSpan(instrumentationName, expectedName).withServiceName("foo").start()
```
Comment (0xnm) [inline features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/CoreSpanBuilderTest.kt:165]:
```suggestion
        val secondSpan = tracer.buildSpan(instrumentationName, expectedName).withServiceName("foo").start()
```

same suggestion applies to other tests - instead of using `var` and changing the reference, it is better to be more explicit in saying that it is another instance.

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/CoreSpanBuilderTest.kt [hunk 5]
```diff
@@ -0,0 +1,534 @@
...
+        val stop = System.currentTimeMillis()
+
+        // Give a range of +/- 5 millis
+        assertThat(span.startTime).isGreaterThanOrEqualTo(TimeUnit.MILLISECONDS.toNanos(start - 1))
+        assertThat(span.startTime).isLessThanOrEqualTo(TimeUnit.MILLISECONDS.toNanos(stop + 1))
```
Comment (0xnm) [inline features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/CoreSpanBuilderTest.kt:170]: actual offset is 1 millisecond, not 5 as it is said in the comment

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/CoreSpanBuilderTest.kt [hunk 6]
```diff
@@ -0,0 +1,534 @@
...
+    fun `TagContext should populate default span details`(tagContext: TagContext) {
+        // setup:
+        val thread = Thread.currentThread()
+        val span = tracer.buildSpan(instrumentationName, "op name").asChildOf(tagContext).start() as DDSpan
+
+        // expect:
+        assertThat(span.traceId).isNotEqualTo(DDTraceId.ZERO)
+        assertThat(span.parentId).isEqualTo(DDSpanId.ZERO)
+        val samplingPriority = span.samplingPriority
+        assert(samplingPriority == null)
```
Comment (0xnm) [inline features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/CoreSpanBuilderTest.kt:353]:
```suggestion
        assertThat(samplingPriority).isNull()
```

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/CoreSpanBuilderTest.kt [hunk 7]
```diff
@@ -0,0 +1,534 @@
...
+                    )
+        )
+    }
+
+    @ParameterizedTest
+    @MethodSource("tagStringProvider")
+    fun `default span tags populated on each span`(tags: Map<String, String>) {
+        // setup:
+        val customTracer = tracerBuilder().writer(writer).config(Config.get()).defaultSpanTags(tags).build()
+        val span = customTracer.buildSpan(instrumentationName, "op name").withServiceName("foo").start()
```
Comment (0xnm) [inline features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/CoreSpanBuilderTest.kt:375]: it is better to wrap such lines call-per-call (and in other tests as well)

**Review by 0xnm — CHANGES_REQUESTED** [review]
I gave it a quick look and added some comments. Also my concern is that some things are public now, can we review if all of them are needed to be public for the sake of the tests? Maybe it is better to access them with reflection instead of making public for everyone?

File: features/dd-sdk-android-trace/src/main/java/com/datadog/trace/common/writer/ListWriter.java [hunk 8]
```diff
@@ -37,6 +37,9 @@ public boolean accept(List<DDSpan> trace) {
   public List<DDSpan> firstTrace() {
     return get(0);
   }
+  public List<DDSpan> secondTrace() {
```
[the remark this answers is not in the document]
Comment (mariusc83) [reply features/dd-sdk-android-trace/src/main/java/com/datadog/trace/common/writer/ListWriter.java:40]: yup, they are using groovy in their tests and they are able to access these private properties directly.

File: features/dd-sdk-android-trace/src/main/java/com/datadog/trace/common/writer/TraceStructureWriter.java [hunk 9]
```diff
@@ -120,7 +120,6 @@ public void write(List<DDSpan> trace) {
                   + ")"
                   + "->"
                   + span.getTraceId();
-          out.println("ERROR: " + message);
```
[the remark this answers is not in the document]
Comment (mariusc83) [reply features/dd-sdk-android-trace/src/main/java/com/datadog/trace/common/writer/TraceStructureWriter.java:123]: too much noise and also if you check about this line the log is added to the `logger` instance for which I am adding the implementation in the next PR to delegate to our internal logger.

File: features/dd-sdk-android-trace/src/main/java/com/datadog/trace/core/propagation/W3CHttpCodec.java [hunk 10]
```diff
@@ -28,347 +28,348 @@
...
+            if (length < TRACE_PARENT_LENGTH) {
+                throw new IllegalStateException("The length of traceparent '" + tp + "' is too short");
+            }
+            long version = LongStringUtils.parseUnsignedLongHex(tp, 0, 2, true);
+            if (version == 255) {
+                throw new IllegalStateException("Illegal version number " + tp.substring(0, 2));
+            } else if (version == 0 && length > TRACE_PARENT_LENGTH) {
+                throw new IllegalStateException("The length of traceparent '" + tp + "' is too long");
+            }
+            DDTraceId traceId = DD128bTraceId.fromHex(tp, TRACE_PARENT_TID_START, 32, true);
```
[the remark this answers is not in the document]
Comment (mariusc83) [reply features/dd-sdk-android-trace/src/main/java/com/datadog/trace/core/propagation/W3CHttpCodec.java:298]: it is going to come later in the next PRs according with Ganesh but Id id not do anything here, this class was modified just because of the format change.

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/DDSpecification.kt [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+            if (keyAsString.lowercase().startsWith("dd_")) {
+                System.getenv(keyAsString)?.let { keyValue -> removedEnvProperties.put(keyAsString, keyValue) }
+            }
+        }
+        removedProperties.keys.forEach { System.clearProperty(it) }
+        removedEnvProperties.keys.forEach { environmentVariables.remove(it) }
+    }
+
+    private fun restoreSystemProperties() {
+        removedProperties.forEach { (key, value) -> System.setProperty(key, value) }
```
[the remark this answers is not in the document]
Comment (mariusc83) [reply features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/DDSpecification.kt:57]: we are using a mock there which in the `StubExtension` so what the mock is doing under the hood is creating a `mock` System.env , adding the original properties to it and this is removed on the `tearDown` so it is done automatically.

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/CoreSpanBuilderTest.kt (hunk 2 above)
Comment (mariusc83) [reply features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/CoreSpanBuilderTest.kt:61]: while I do agree on the other hand this will take a lot of time to cleanup all these tests and add randomness. I am not sure we should invest this time here now as by the end of the year we might have to drop this code anyway. @plousada ?

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/CoreSpanBuilderTest.kt [hunk 12]
```diff
@@ -0,0 +1,534 @@
...
+        assertThat(context.errorFlag).isTrue
+        assertThat(context.serviceName).isEqualTo(expectedService)
+        assertThat(context.spanType).isEqualTo(expectedType)
+
+        assertThat(context.getTag(DDTags.THREAD_NAME)).isEqualTo(Thread.currentThread().name)
+        assertThat(context.getTag(DDTags.THREAD_ID)).isEqualTo(Thread.currentThread().id)
+    }
+
+    @Test
+    fun `setting name should remove`() {
```
[the remark this answers is not in the document]
Comment (mariusc83) [reply features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/CoreSpanBuilderTest.kt:124]: yeah...this was copy pasted from their tests, honestly did not have the time to check all the function names but will try to figure out what to add there :)

File: features/dd-sdk-android-trace/api/apiSurface
[the remark this answers is not in the document]
Comment (mariusc83) [reply features/dd-sdk-android-trace/api/apiSurface:1]: why restored ? these are api changes reflected by the changes we added in the java project to make some classes/properties available for our unit tests.

File: features/dd-sdk-android-trace/src/main/java/com/datadog/trace/api/Config.java [hunk 13]
```diff
@@ -3800,7 +3800,7 @@ private static String getProp(String name, String def) {
 
     // This has to be placed after all other static fields to give them a chance to initialize
 
-    private static final Config INSTANCE =
```
[the remark this answers is not in the document]
Comment (0xnm) [reply features/dd-sdk-android-trace/src/main/java/com/datadog/trace/api/Config.java:3803]: yes, we have tests which set `static final` on JDK 17 :)

File: features/dd-sdk-android-trace/src/main/java/com/datadog/trace/common/writer/ListWriter.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (0xnm) [reply features/dd-sdk-android-trace/src/main/java/com/datadog/trace/common/writer/ListWriter.java:40]: let's maybe do the same? write reflection wrappers to access all this. Because otherwise we make quite a lot of internal things public just for the tests, not sure if it opens the door for the misuse on the client side.

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/DDSpecification.kt (hunk 11 above)
[the remark this answers is not in the document]
Comment (0xnm) [reply features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/DDSpecification.kt:57]: but then what is the point to record values into `removedEnvProperties` if they are never going to be accessed in order to be restored?

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/CoreSpanBuilderTest.kt (hunk 2 above)
Comment (0xnm) [reply features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/CoreSpanBuilderTest.kt:61]: I don't think it takes a lot of time, but it greatly impacts the readability of the PR, because currently everything is mixed and it is hard to follow test body content tbh.

File: features/dd-sdk-android-trace/api/apiSurface
[the remark this answers is not in the document]
Comment (0xnm) [reply features/dd-sdk-android-trace/api/apiSurface:1]: because this file is deleted completely in this PR. It should still be there with the actual Kotlin API of this module.

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/DDSpecification.kt (hunk 11 above)
[the remark this answers is not in the document]
Comment (mariusc83) [reply features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/DDSpecification.kt:57]: I actually need to just use a private variable there.

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/CoreSpanBuilderTest.kt (hunk 2 above)
Comment (mariusc83) [reply features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/CoreSpanBuilderTest.kt:61]: I will take care of the `given, when, then` for now

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/CoreSpanBuilderTest.kt (hunk 6 above)
Comment (mariusc83) [reply features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/CoreSpanBuilderTest.kt:353]: yeah for some reason that is not working, failing with a NPE and I think is because of the Java function signature at the end of the line.

File: features/dd-sdk-android-trace/src/main/java/com/datadog/trace/common/writer/ListWriter.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (mariusc83) [reply features/dd-sdk-android-trace/src/main/java/com/datadog/trace/common/writer/ListWriter.java:40]: it shouldn't honestly, there are plenty of places like these in their api's.

[the remark this answers is not in the document]
Comment (0xnm) [reply features/dd-sdk-android-trace/src/main/java/com/datadog/trace/common/writer/ListWriter.java:40]: I think it makes sense if you think about this code as an agent - it is attached to the process and is not supposed to be used as a SDK.

But in our case it becomes a SDK, and we should care much more about things which should be open vs closed.

[the remark this answers is not in the document]
Comment (mariusc83) [reply features/dd-sdk-android-trace/src/main/java/com/datadog/trace/common/writer/ListWriter.java:40]: actually this class is only used for testing purposes on their end, it blocks the thread so to make sure the traces are consumed. I think this one is safe.

**Review by 0xnm — COMMENTED** [review]
nice work! I added few comments/suggestions, but nothing blocking.

File: .gitlab-ci.yml [hunk 14]
```diff
@@ -16,6 +16,7 @@ variables:
 
   KUBERNETES_MEMORY_REQUEST: "8Gi"
   KUBERNETES_MEMORY_LIMIT: "16Gi"
+  DD_TRACE_REMOVE_INTEGRATION_SERVICE_NAMES_ENABLED: "false"
```
[the remark this answers is not in the document]
Comment (mariusc83) [reply .gitlab-ci.yml:19]: it is not affecting any other tests, it is strictly related with the tracing tests and code but yes I could add it there, I just did not want to duplicate this line.

File: features/dd-sdk-android-trace/src/main/java/com/datadog/trace/core/CoreTracer.java [hunk 15]
```diff
@@ -264,7 +265,7 @@ public static class CoreTracerBuilder {
 
         private Config config;
         private String serviceName;
-        private Writer writer;
+        private Writer writer = new NoOpWriter();
```
[the remark this answers is not in the document]
Comment (mariusc83) [reply features/dd-sdk-android-trace/src/main/java/com/datadog/trace/core/CoreTracer.java:268]: because otherwise you cannot build a tracer without passing a writer and in case you do it will throw a NoOp. Before the builder had a default value but I removed that when cleaning up the code and decided to have it as nullable. During these unit tests refactor I realized that is better to have it as NoOp in case it is not provided to remove the noise in the tests.

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/PendingTraceTest.kt [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+        val trace = PendingTrace(tracer, DDTraceId.from(0), buffer, mock(), null, false, healthMetrics)
+
+        // Then
+        whenever(buffer.longRunningSpansEnabled()).thenReturn(true)
+
+        // When
+        val span1 = createSimpleSpanWithID(trace, 39)
+        span1.setFieldValue("durationNano", 31)
+        @Suppress("DEPRECATION")
+        span1.samplingPriority = PrioritySampling.USER_KEEP.toInt()
```
[the remark this answers is not in the document]
Comment (mariusc83) [reply features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/PendingTraceTest.kt:101]: I will double check

File: tools/unit/src/main/kotlin/com/datadog/tools/unit/ReflectUtils.kt [hunk 17]
```diff
@@ -114,3 +116,95 @@ private fun <R> setFieldValue(instance: Any?, field: Field, fieldValue: R): Bool
...
+    val output: Any?
+    method.isAccessible = true
+    try {
+        output = if (params.isEmpty()) {
+            method.invoke(null)
+        } else {
+            method.invoke(null, *params)
+        }
+    } catch (e: InvocationTargetException) {
+        throw e.cause ?: e
```
[the remark this answers is not in the document]
Comment (mariusc83) [reply tools/unit/src/main/kotlin/com/datadog/tools/unit/ReflectUtils.kt:158]: I just followed what was done in the `ReflectUtils` in the nightly tests module. I supposed we had a good reason to do that there probably to get the underlying cause for better visibility ?

[the remark this answers is not in the document]
Comment (0xnm) [reply tools/unit/src/main/kotlin/com/datadog/tools/unit/ReflectUtils.kt:158]: ok, makes sense. it doesn't impact anything anyway, I was rather curious about the reasoning.

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/PendingTraceTestBase.kt [hunk 18]
```diff
@@ -0,0 +1,291 @@
...
+    @Test
+    fun `test getCurrentTimeNano`() {
+        // Generous 5 seconds to execute this test
+        assertThat(
+            Math.abs(
+                TimeUnit.NANOSECONDS.toSeconds(trace.currentTimeNano) -
+                        TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis())
+            )
+        ).isLessThan(5)
+    }
```
[the remark this answers is not in the document]
Comment (mariusc83) [reply features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/PendingTraceTestBase.kt:142]: not sure, this is what they have in their test classes, I think they want to assert that the clock is not skewed

File: tools/unit/src/main/kotlin/com/datadog/tools/unit/ReflectUtils.kt [hunk 19]
```diff
@@ -114,3 +116,95 @@ private fun <R> setFieldValue(instance: Any?, field: Field, fieldValue: R): Bool
...
+@Suppress("SpreadOperator")
+inline fun <reified T : Any, reified R> Class<T>.callStaticMethod(
+    methodName: String,
+    vararg params: Any?
+): R {
+    val declarationParams = Array<Class<*>?>(params.size) {
+        params[it]?.javaClass ?: Any::class.java
+    }
+
+    val method = this.getDeclaredMethodRecursively(methodName, true, declarationParams)
```
[the remark this answers is not in the document]
Comment (mariusc83) [reply tools/unit/src/main/kotlin/com/datadog/tools/unit/ReflectUtils.kt:146]: yeah you are right it is not needed here.

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/DDSpanTest.kt [hunk 20]
```diff
@@ -0,0 +1,588 @@
...
+            .asChildOf(extractedContext)
+            .start()
+        val child = tracer
+            .buildSpan(instrumentationName, "child")
+            .asChildOf(root.context())
+            .start()
+
+        // Then
+        assertThat(child.localRootSpan).isEqualTo(root)
+        assertThat(child.localRootSpan).isEqualTo(root)
```
[the remark this answers is not in the document]
Comment (mariusc83) [reply features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/DDSpanTest.kt:358]: yeah I think they copied the line there, I will double check

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/DDSpanTest.kt [hunk 21]
```diff
@@ -0,0 +1,588 @@
...
+            .buildSpan(instrumentationName, "testChild1")
+            .asChildOf(parent)
+            .start()
+            .context() as DDSpanContext
+
+        // Then
+        assertThat(child.origin).isEqualTo("some-origin")
+        // Access field directly instead of getter not directly possible in Kotlin
+        assertThat(child.origin).isEqualTo("some-origin")
+        // Access field directly instead of getter not directly possible in Kotlin, Nullable origin would be null at runtime
```
[the remark this answers is not in the document]
Comment (mariusc83) [reply features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/DDSpanTest.kt:316]: oops, nope

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/DDSpanContextTest.kt [hunk 22]
```diff
@@ -0,0 +1,346 @@
...
+
+        // Then
+        val thread = Thread.currentThread()
+        val expectedTags = mapOf(DDTags.THREAD_NAME to thread.name, DDTags.THREAD_ID to thread.id)
+        assertThat(context.tags).containsAllEntriesOf(expectedTags)
+        assertThat(
+        context::class.java.getMethod(method)
+            .invoke(context)
+        )
+            .isEqualTo(value)
```
[the remark this answers is not in the document]
Comment (mariusc83) [reply features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/DDSpanContextTest.kt:85]: yeah because I do no have the changes in develop branch where you enabled the formatting yet.

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/PendingTraceTestBase.kt (hunk 18 above)
[the remark this answers is not in the document]
Comment (0xnm) [reply features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/PendingTraceTestBase.kt:142]: there is non-system clock in the code we are porting? is it using NTP then or what if it is the case?

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/CoreTracerTest.kt [hunk 23]
```diff
@@ -0,0 +1,235 @@
...
+        val writer = ListWriter()
+        val tracer = tracerBuilder().writer(writer).build()
+
+        // When
+        val root = tracer.buildSpan(instrumentationName, "operation").start()
+        val child = tracer.buildSpan(instrumentationName, "my_child").asChildOf(root.context()).start()
+        root.finish()
+
+        // Then
+        assert(child.samplingPriority == null)
```
[the remark this answers is not in the document]
Comment (mariusc83) [reply features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/CoreTracerTest.kt:198]: Java code returns NULL instead of Int? and that code crashes with an NPE, I could cast it but I prefer it this way. Same for the other comment on the same topic.

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/PendingTraceTest.kt (hunk 16 above)
[the remark this answers is not in the document]
Comment (mariusc83) [reply features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/PendingTraceTest.kt:101]: In their tests they are using the same class so I would rather keep it but will double check.

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/PendingTraceTestBase.kt (hunk 18 above)
[the remark this answers is not in the document]
Comment (mariusc83) [reply features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/PendingTraceTestBase.kt:142]: I am really not sure, I really cannot without asking them and I cannot really ask all these details. We have 2 options here, decide to keep all the tests they have or we drop some.

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/CoreTracerTest.kt (hunk 23 above)
[the remark this answers is not in the document]
Comment (0xnm) [reply features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/CoreTracerTest.kt:198]: Is it possible to use `check` then if using `assertThat` brings the problem here (not sure though still where exactly the problem lies)?

The thing this line will alway pass unless JVM is started with `-ea` flag (enable assertions), because this `assert(...)` is shortcut for the Java's `assert ....` (https://docs.oracle.com/javase/8/docs/technotes/guides/language/assert.html).

Event Kotlin's `assert` docs says:

> Throws an [AssertionError] if the [value] is false and runtime assertions have been enabled on the JVM using the *-ea* JVM option.

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/PendingTraceTest.kt (hunk 16 above)
[the remark this answers is not in the document]
Comment (0xnm) [reply features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/PendingTraceTest.kt:101]: well, this is essentially the same class in 2 different places of the codebase. Anyway, not sure what is the point to use the deprecated one when there is non-deprecated.

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/PendingTraceTestBase.kt (hunk 18 above)
[the remark this answers is not in the document]
Comment (0xnm) [reply features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/PendingTraceTestBase.kt:142]: ok, let's keep it then. If there is a NTP support, then it is probably some code around it/network stack while we have our own NTP support (so double the code for the same).

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/CoreTracerTest.kt (hunk 23 above)
[the remark this answers is not in the document]
Comment (mariusc83) [reply features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/CoreTracerTest.kt:198]: actually was an issue in the Java code, that method was not marked as @Nullable and Kotlin was not returning an Int?

File: features/dd-sdk-android-trace/api/dd-sdk-android-trace.api [hunk 24]
```diff
@@ -6034,6 +6049,7 @@ public class com/datadog/trace/core/CoreTracer$CoreTracerBuilder {
 
 public class com/datadog/trace/core/DDSpan : com/datadog/trace/api/profiling/TransientProfilingContextHolder, com/datadog/trace/bootstrap/instrumentation/api/AgentSpan, com/datadog/trace/bootstrap/instrumentation/api/AttachableWrapper, com/datadog/trace/core/CoreSpan {
 	public static final field CHECKPOINTED_TAG Ljava/lang/String;
+	public fun <init> (Ljava/lang/String;JLcom/datadog/trace/core/DDSpanContext;Ljava/util/List;)V
```
Comment (0xnm) [inline features/dd-sdk-android-trace/api/dd-sdk-android-trace.api:6052]: minor: given that we have a method to call private constructor now, should we make this constructor public?

File: .gitlab-ci.yml [hunk 25]
```diff
@@ -132,6 +132,53 @@ test:release:
...
+# We need separated steps for tracing module tests as these cannot be run with the CI visibility agent enabled. Also they need to be executed with
+# the daemon capability as the worker threads are running as a daemon.
+
+test:trace-debug:
+  tags: [ "arch:amd64" ]
+  image: $CI_IMAGE_DOCKER
+  stage: test
+  timeout: 1h
+  variables:
+    DD_TRACE_REMOVE_INTEGRATION_SERVICE_NAMES_ENABLED: "false"
```
Comment (0xnm) [inline .gitlab-ci.yml:144]: btw is this flag still needed given that we clean up env variable in the `DDSpecification` and default value of this one in `Config` is `false`?

**Review by 0xnm — APPROVED** [review]
lgtm. I left few questions, but nothing blocking.

Also speaking of https://github.com/DataDog/dd-sdk-android/pull/1924#discussion_r1539406052 - should we add nullable/non-nullable annotations to the public Java API we bringing to have a better interop with Kotlin in order to avoid such things on the customer side? Or the only API exposed / advised to be used will be in Kotlin?

File: .gitlab-ci.yml (hunk 25 above)
Comment (mariusc83) [reply .gitlab-ci.yml:144]: yes I am not sure where exactly before that is still used...before the Config is even created. as you could see in my tests I assert that there is no more any DD_ property in any System or System.env and still this poses problems. Alessio from APM team recommended me to add it like this to make sure.

File: features/dd-sdk-android-trace/api/dd-sdk-android-trace.api (hunk 24 above)
Comment (mariusc83) [reply features/dd-sdk-android-trace/api/dd-sdk-android-trace.api:6052]: actually not...

**Comment by mariusc83** [comment]
> lgtm. I left few questions, but nothing blocking.
> 
> Also speaking of [#1924 (comment)](https://github.com/DataDog/dd-sdk-android/pull/1924#discussion_r1539406052) - should we add nullable/non-nullable annotations to the public Java API we bringing to have a better interop with Kotlin in order to avoid such things on the customer side? Or the only API exposed / advised to be used will be in Kotlin?

We only want to provide Kotlin support so normally they should not touch the Java API but it might be worthy to have the discussion for any other Java support we offer, outside of Trace.

File: .gitlab-ci.yml (hunk 25 above)
Comment (0xnm) [reply .gitlab-ci.yml:144]: We can remove it and see if it is still needed. Maybe it is not anymore.

**Comment by 0xnm** [comment]
@mariusc83 Cannot it be the case that some Java objects (like spans, etc.) are exposed in the Kotlin APIs we are going to provide?

File: features/dd-sdk-android-trace/api/dd-sdk-android-trace.api (hunk 24 above)
Comment (mariusc83) [reply features/dd-sdk-android-trace/api/dd-sdk-android-trace.api:6052]: I am not really sure where this is coming from as I am just exposing the class not the constructors so I will leave it as is. I need the class exposed to cast to `DDSpan` in the unit tests.

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/PendingTraceTest.kt (hunk 16 above)
[the remark this answers is not in the document]
Comment (mariusc83) [reply features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/PendingTraceTest.kt]: yes, I know, I will probably figure it out better later.

**Comment by mariusc83** [comment]
> @mariusc83 Cannot it be the case that some Java objects (like spans, etc.) are exposed in the Kotlin APIs we are going to provide?

The idea about OTel support is to only use the OTel API so they should not see any Java objects in their code. I think we are safe here.

File: features/dd-sdk-android-trace/api/dd-sdk-android-trace.api (hunk 24 above)
Comment (0xnm) [reply features/dd-sdk-android-trace/api/dd-sdk-android-trace.api:6052]: It is from this change https://github.com/DataDog/dd-sdk-android/pull/1924/files#diff-1c12b21f74915bf39d879d5d17374ad221d56bb67bed99743adc8a86dfbf944cR127 where private constructor was made public. But that is ok I believe, just was wondering about the consistency in the approach (since for `PendingTrace` we are using reflection to create an instance).

**Review by 0xnm — APPROVED** [review]
